core/include/vecmem/memory/details/hip_device_atomic_ref.hpp

```diff
@@ -0,0 +1,33 @@
+/*
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2022-2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+// HIP include(s).
+#include <hip/hip_runtime.h>
+
+// Set up __VECMEM_THREADFENCE correctly for the vecmem::hip::device_atomic_ref
+// code.
+#ifdef __HIP_DEVICE_COMPILE__
+#define __VECMEM_THREADFENCE __threadfence()
+#else
+#define __VECMEM_THREADFENCE
+#endif  // defined(__HIP_DEVICE_COMPILE__)
```

Comment:
I thought I would be able to avoid all such tricky logic with this code layout, but as it turns out, no...

You see, once I made `vecmem::posix_device_atomic_ref` only work in host code, it turned out that at least HIP (but I think also CUDA...) was instantiating that type for at least some steps in building `__global__` functions.

But once I made sure that the CUDA and HIP compilers would always see `vecmem::cuda::device_atomic_ref` and `vecmem::hip::device_atomic_ref`, even when building host code, I ran into complaints from `__threadfence()` being declared a "device only" function, so we're not allowed to call it from a `VECMEM_HOST_AND_DEVICE` function.

When hiding the `__threadfence()` call behind `__CUDA_ARCH__` and `__HIP_DEVICE_COMPILE__`, this compiler complaint disappears. Apparently `nvcc` doesn't check for this type of consistency when only doing the device code compilation stage of the build. :confused:

Long story short, apparently this part of our code has to stay inherently fragile after all.
  - As long as we want to allow `vecmem::device_vector` to work in host code (and we absolutely do!), `vecmem::cuda::device_atomic_ref` has to use `VECMEM_HOST_AND_DEVICE` on its functions.
  - `__threadfence()` is a "device only" function, so calling it from a `VECMEM_HOST_AND_DEVICE` function is not actually allowed.

So yeah, I don't see a clean way out of this fragility at the moment. Not without exposing a completely different class during the host and device code compilation stages to `nvcc`. Which is also not any less fragile... :frowning:

Reply:
Switching (back) to `__CUDA_ARCH__` and `__HIP_DEVICE_COMPILE__` made this macro unnecessary once again... 🤔